sqld/src/lib.rs

```diff
@@ -144,87 +225,29 @@ pub async fn run_server(config: Config) -> anyhow::Result<()> {
         ))
     });
 
-    // create database directory where all the files will go.
-    if !config.db_path.exists() {
-        std::fs::create_dir_all(&config.db_path)?;
-    }
+    loop {
+        if !config.db_path.exists() {
+            std::fs::create_dir_all(&config.db_path)?;
+        }
+        let mut handles = FuturesUnordered::new();
 
-    match config.writer_rpc_addr {
-        Some(ref addr) => {
-            let factory = WriteProxyDbFactory::new(
-                addr,
-                config.writer_rpc_tls,
-                config.writer_rpc_cert.clone(),
-                config.writer_rpc_key.clone(),
-                config.writer_rpc_ca_cert.clone(),
-                config.db_path.clone(),
-                #[cfg(feature = "mwal_backend")]
-                vwal_methods,
-            )
-            .await
-            .context("failed to start WriteProxy DB")?;
-            let service = DbFactoryService::new(factory);
-            run_service(service, config).await?;
+        match config.writer_rpc_addr {
+            Some(ref addr) => start_primary(&config, &mut handles, addr).await?,
```

Comment:
wouldn't it be cleaner for start_primary and start_replica to return a handle that this function adds to handles?

Reply:
Or collection of handles if there's more than one.